tests/it/compute/temporal.rs

```diff
@@ -2,173 +2,317 @@ use arrow2::array::*;
 use arrow2::compute::temporal::*;
 use arrow2::datatypes::*;
 
-#[test]
-fn date64_hour() {
-    let array =
-        Int64Array::from(&[Some(1514764800000), None, Some(1550636625000)]).to(DataType::Date64);
+macro_rules! temporal_test {
+    ($func:ident, $extract:ident, $data_types:path) => {
+        #[test]
+        fn $func() {
+            for data_type in $data_types() {
+                let data = TestData::data(&data_type);
+                let result = $extract(&*data.input).unwrap();
 
-    let result = hour(&array).unwrap();
-    let expected = UInt32Array::from(&[Some(0), None, Some(4)]);
-    assert_eq!(result, expected);
+                assert_eq!(
+                    result,
+                    data.$extract.unwrap(),
+                    "\"{}\" failed on type: {:?}",
+                    stringify!($extract),
+                    data_type
+                );
+            }
+        }
+    };
 }
 
-#[test]
-fn date32_hour() {
-    let array = Int32Array::from(&[Some(15147), None, Some(15148)]).to(DataType::Date32);
+temporal_test!(temporal_hour, hour, TestData::available_time_like_types);
+temporal_test!(temporal_minute, minute, TestData::available_time_like_types);
+temporal_test!(temporal_second, second, TestData::available_time_like_types);
+temporal_test!(
+    temporal_nanosecond,
+    nanosecond,
+    TestData::available_time_like_types
+);
+temporal_test!(temporal_year, year, TestData::available_date_like_types);
+temporal_test!(temporal_month, month, TestData::available_date_like_types);
+temporal_test!(temporal_day, day, TestData::available_date_like_types);
+temporal_test!(
+    temporal_weekday,
+    weekday,
+    TestData::available_date_like_types
+);
+temporal_test!(
+    temporal_iso_week,
+    iso_week,
+    TestData::available_date_like_types
+);
 
-    let result = hour(&array).unwrap();
-    let expected = UInt32Array::from(&[Some(0), None, Some(0)]);
-    assert_eq!(result, expected);
+struct TestData {
+    input: Box<dyn Array>,
+    year: Option<Int32Array>,
+    month: Option<UInt32Array>,
+    day: Option<UInt32Array>,
+    weekday: Option<UInt32Array>,
+    iso_week: Option<UInt32Array>,
+    hour: Option<UInt32Array>,
+    minute: Option<UInt32Array>,
+    second: Option<UInt32Array>,
+    nanosecond: Option<UInt32Array>,
 }
 
-#[test]
-fn time32_second_hour() {
-    let array = Int32Array::from(&[Some(37800), None]).to(DataType::Time32(TimeUnit::Second));
+impl TestData {
+    fn data(data_type: &DataType) -> TestData {
+        match data_type {
+            DataType::Date64 => TestData {
+                input: Box::new(
+                    Int64Array::from(&[Some(1514764800000), None, Some(1550636625000)])
```

Comment:
The third entry in this test might be a little misleading given 86400000 constraint on `date64`